<a href="https://colab.research.google.com/github/tiggofreire/estrutura-de-dados-II/blob/main/Atividade-aula03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exercício 1**
Execute:

In [1]:
a = [10, 20, 30]
b = a

b.append(40)

print("a:", a)
print("b:", b)
print("id(a):", id(a))
print("id(b):", id(b))

a: [10, 20, 30, 40]
b: [10, 20, 30, 40]
id(a): 132052177454016
id(b): 132052177454016


Responda:

- Qual será a saída?
 R: a: [10, 20, 30, 40]
b: [10, 20, 30, 40]
id(a): 132052177454016
id(b): 132052177454016
- a e b representam o mesmo objeto?
si, as duas variáveis levamm pra mesma lista na memória. Por isso, id(a) e id(b) são iguais.
- Por quê?
Porque b = a não cria uma nova lista. Ele só faz b receber uma referência para o mesmo objeto da referencia. Aí quando fazemos b.append(40), a lista original é modificada, e a alteração aparece  em a e b.
- O que aconteceria com b = a.copy()?
Os ids também seriam diferentes, pois a e b seriam objetos diferentes.




# **Exercício 2 — Construindo uma cadeia**
Crie:

In [ ]:
n1 = Node("A")
n2 = Node("B")
n3 = Node("C")

Conecte:

In [ ]:
A → B → C → None

Depois percorra a estrutura e produza:

In [ ]:
A
B
C

In [6]:
#Resposta

class Node:
    def __init__(self, valor):
        self.valor = valor
        self.proximo = None


n1 = Node("A")
n2 = Node("B")
n3 = Node("C")

# Conectando os nós
n1.proximo = n2
n2.proximo = n3
n3.proximo = None

# Percorrendo a estrutura
atual = n1

while atual is not None:
    print(atual.valor)
    atual = atual.proximo


A
B
C


# **Exercício 3 — Depuração**
Analise:

In [ ]:
n1 = Node(10)
n2 = Node(20)
n3 = Node(30)

n1.proximo = n2
n2.proximo = n3

atual = n1

while atual is not None:
    print(atual.valor)

Perguntas:

- Qual é o problema?
a variável atual nunca é atualizada.
- Por que o programa não termina?
Porque atual continua apontando para n1 imprimindo 10 eternamente.
- Qual linha deve ser acrescentada?
algo como atual = atual.proximo

- Qual será a saída depois da correção?
10
20
30


# **Problema real — Sistema de atendimento de uma clínica**
Uma clínica recebe pacientes durante o dia. Cada paciente possui:

- nome;
- idade;
- prioridade.
- Exemplo:

In [ ]:
Ana — 32 anos — Normal
Bruno — 70 anos — Prioridade
Carlos — 45 anos — Normal

O sistema deve permitir:

1. adicionar paciente;
2. listar pacientes;
3. atender o primeiro paciente;
4. verificar se a fila está vazia;
5. informar a quantidade de pacientes.